src/or/voting_schedule.c

```diff
@@ -138,14 +142,32 @@ voting_schedule_t voting_schedule;
 time_t
 voting_schedule_get_next_valid_after_time(void)
 {
+  bool need_to_recalculate_voting_schedule = false;
+
   /* This is a safe guard in order to make sure that the voting schedule
    * static object is at least initialized. Using this function with a zeroed
    * voting schedule can lead to bugs. */
   if (tor_mem_is_zero((const char *) &voting_schedule,
                       sizeof(voting_schedule))) {
-    voting_schedule_recalculate_timing(get_options(), time(NULL));
+    need_to_recalculate_voting_schedule = true;
+  }
+
+  /* Also make sure we are not using an outdated voting schedule. If we have a
+   * newer consensus, make sure we recalculate the voting schedule. */
+  networkstatus_t *consensus = networkstatus_get_live_consensus(approx_time());
+  if (consensus &&
+      consensus->valid_after != voting_schedule.live_consensus_valid_after) {
+    log_info(LD_DIR, "Voting schedule is outdated: recalculating (%d/%d)",
+             (int) consensus->valid_after,
+             (int) voting_schedule.live_consensus_valid_after);
+    need_to_recalculate_voting_schedule = true;
+  }
+
+  if (need_to_recalculate_voting_schedule) {
+    voting_schedule_recalculate_timing(get_options(), approx_time());
```

Comment:
Hold on a sec. Don't we have a circular loop here? `get_voting_schedule()` can reach this point and then `voting_schedule_recalculate_timing()` calls `get_voting_schedule()` again so we do this whole path over and over again?

I actually wonder if the right approach would be to have a getter for the `voting_schedule` static object that makes sure it is not out of date. So no where we should use that object directly but rather get the ref through the getter assuring us that it is updated?

Reply:
Hmmm, trying to understand this better: How can `get_voting_schedule()` reach `voting_schedule_get_next_valid_after_time()`? They seem disconnected to me.

No?

Reply:
Oh oops... my brain failed to parse that `voting_schedule_get_start_of_next_interval()` (in get_voting_schedule()) `!= voting_schedule_get_next_valid_after_time()` :P...

NVM.